graphql.go

```diff
@@ -118,6 +119,36 @@ var (
 	}
 )
 
+// Returns a request ptr that is of same value with input request
+func (c *Client) cloneRequest(req *http.Request) (*http.Request, error) {
+	// Read request body until the end
+	bodyBytes, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading request body: %s", err)
+	}
+
+	// Save the request body back to the request
+	req.Body = ioutil.NopCloser(bytes.NewReader(bodyBytes))
+
+	// Create a completely new request with same body and endpoint
+	newRequest, err := http.NewRequest(http.MethodPost, c.endpoint, bytes.NewBuffer(bodyBytes))
+	if err != nil {
+		return nil, fmt.Errorf("Error creating a new request: %s", err)
+	}
+
+	// Set header
+	for key, values := range req.Header {
+		for _, value := range values {
+			newRequest.Header.Add(key, value)
+		}
+	}
+
+	// Set context
+	newRequest = newRequest.WithContext(req.Context())
+
+	return newRequest, nil
+}
```

Comment:
Instead of cloning the request, you could try something like this to just tee req.Body into a buffer that replaces the request body each time:

```
var body io.Reader = req.Body

for tryCount := 0; tryCount < retryConfig.MaxTries; tryCount++ {
	buf := new(bytes.Buffer)
	req.Body = ioutil.NopCloser(io.TeeReader(body, buf))

	resp, err = c.httpClient.Do(req)
	if err == nil && !retryConfig.shouldRetry(resp.StatusCode) {
		return resp, nil
	}

	body = buf
         ...
}
```